### PR TITLE
switching from mongo to mysql and adding threshold to number of stored git-logs

### PR DESCRIPTION
**Pull request by HamzaIbnFarooq**
#### Pre-Flight checklist
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested

#### What are the relevant tickets?
https://github.com/mitodl/edx-sysadmin/issues/12

#### What's this PR do?
* Switches from Mongo to MySQL for GitLogs
* Add threshold (SYSADMIN_MAX_GIT_LOGS_THRESHOLD) value to remove old GitLogs per course

#### How should this be manually tested?
* Test the whole git import and git log feature to test if MySQL works perfectly with GitLogs.
* Set `SYSADMIN_MAX_GIT_LOGS_THRESHOLD` variable to a specific `integer` through your `lms.yml` or `private.py` in `edx-platform` and then import a course. It should delete the old course logs of that specific course if there are more than the defined threshold.

#### Any background context you want to provide?
The main issue we had with MySQL earlier was that the size of Course-Import Logs could get longer than the allowed TextField size (https://electrictoolbox.com/maximum-length-mysql-text-field-types/) so we had to use MongoDB for storing logs.

Now as we have JSONField in MySQL we can use that to solve the problem and get rid of MongoDB. The size of JSONField in MySQL is directly dependent on `max-allowed-packet` global variable (https://stackoverflow.com/questions/40711101/max-json-column-length-in-mysql) which can range up to 1GB and can be easily changed (https://docs.oracle.com/cd/E19182-01/820-6323/gicxk/index.html). So this PR uses that MySQL JSONField.

### CODE REVIEW

**Comment by HamzaIbnFarooq** [comment]
Putting this PR to `blocked` state `until we get some huge edx course` to test this switching.

**Comment by pdpinch** [comment]
I don’t know how large a course we need for this, but here are some possibilities:

https://github.com/mitocw/content-mit-latex2edx-demo
https://github.com/edx/demo-test-course
https://github.com/edx/demo-performance-course

> Putting this PR to blocked state until we get some huge edx course to test this switching.
>

**Comment by HamzaIbnFarooq** [comment]
> https://github.com/mitocw/content-mit-latex2edx-demo
> https://github.com/edx/demo-test-course
> https://github.com/edx/demo-performance-course

Considering these three courses as a benchmark I can say that the feature is working seemlesly with MySQL.

File: edx_sysadmin/models.py [hunk 1]
```diff
@@ -1,23 +1,19 @@
...
 """
+from django.db import models
+from jsonfield.fields import JSONField
 
-import mongoengine
+from opaque_keys.edx.django.models import CourseKeyField
 
-from xmodule.modulestore.mongoengine_fields import CourseKeyField
 
+class CourseGitLog(models.Model):
```
Comment (HamzaIbnFarooq) [inline edx_sysadmin/models.py:10]: FYI: Here I have removed `location` field as it was never getting used.

File: edx_sysadmin/models.py [hunk 2]
```diff
@@ -1,23 +1,19 @@
...
-    git_log = mongoengine.StringField(max_length=65535)
-    repo_dir = mongoengine.StringField(max_length=128)
-    commit = mongoengine.StringField(max_length=40, null=True)
-    author = mongoengine.StringField(max_length=500, null=True)
-    date = mongoengine.DateTimeField()
-    created = mongoengine.DateTimeField()
-    meta = {"indexes": ["course_id", "created"], "allow_inheritance": False}
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    import_log = JSONField(null=True, blank=True)
+    git_log = models.TextField(null=True, blank=True, max_length=65535)
```
Comment (ziafazal) [inline edx_sysadmin/models.py:15]: I don't think we need to specify `max_length` for TextField

File: edx_sysadmin/models.py [hunk 3]
```diff
@@ -1,23 +1,19 @@
...
-    import_log = mongoengine.StringField(max_length=20 * 65535)
-    git_log = mongoengine.StringField(max_length=65535)
-    repo_dir = mongoengine.StringField(max_length=128)
-    commit = mongoengine.StringField(max_length=40, null=True)
-    author = mongoengine.StringField(max_length=500, null=True)
-    date = mongoengine.DateTimeField()
-    created = mongoengine.DateTimeField()
-    meta = {"indexes": ["course_id", "created"], "allow_inheritance": False}
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    import_log = JSONField(null=True, blank=True)
```
Comment (ziafazal) [inline edx_sysadmin/models.py:14]: what is the difference between `git_log` and and `import_log`?

File: edx_sysadmin/models.py [hunk 4]
```diff
@@ -1,23 +1,19 @@
...
+    git_log = models.TextField(null=True, blank=True, max_length=65535)
+    repo_dir = models.CharField(max_length=255, db_index=True)
+    commit = models.CharField(max_length=40, null=True)
+    author = models.CharField(max_length=255, db_index=True)
```
Comment (ziafazal) [inline edx_sysadmin/models.py:18]: too many indexes? do we need index on `author` and `repo_dir` fields?

**Review by ziafazal — COMMENTED** [review]
@HamzaIbnFarooq overall looks good. I have couple of minor suggestions and a question.

File: edx_sysadmin/models.py (hunk 2 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/models.py]: yes, we can remove that because 65535 is the max length already. Updated.

File: edx_sysadmin/models.py (hunk 3 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/models.py]: These two names were used in the old mongo schema where `import_log` stores the `logs produced while importing the course into edx` (now I have changed its name to `course_import_log`). On the other hand, `git_log` stores the git details like `Cloning into 'demo-test-course'... Commit ID: e8458e3631c8b9332fdf9cda2bb46b7c464ad50a Branch: master` (now I have changed its name to `git_details`)

File: edx_sysadmin/models.py (hunk 4 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/models.py]: Thanks for pointing this out, those were added to imitate the previous schema but right now we are fetching `git-logs` on the basis of `course_id` only so I have removed those indexing now.

File: edx_sysadmin/models.py (hunk 3 above)
Comment (ziafazal) [reply edx_sysadmin/models.py]: @HamzaIbnFarooq `git_log` seems more relevant can we revert it?

**Comment by ziafazal** [comment]
@HamzaIbnFarooq can we remove `mongoengine` from requirements and tox configuration?

**Comment by HamzaIbnFarooq** [comment]
> @HamzaIbnFarooq can we remove `mongoengine` from requirements and tox configuration?

@ziafazal I have removed monogengine throughout application.

File: edx_sysadmin/models.py (hunk 3 above)
Comment (HamzaIbnFarooq) [reply edx_sysadmin/models.py]: reverted back.

**Review by ziafazal — APPROVED** [review]
👍

**Comment by pdpinch** [comment]
Can someone from @mitodl/devops take a look at this too?

**Comment by blarghmatey** [comment]
In general the code looks fine, but I'd like to have some functionality to limit the number of log records that are stored in the database per course so that the disk usage doesn't explode with these large log files. I would also like for the limit on the number of records to be configurable via the lms.yml file.

**Comment by HamzaIbnFarooq** [comment]
@blarghmatey I have added a threshold that will remove older logs, kindly have a look.

Besides that, do we need any management command as well for this feature?

File: edx_sysadmin/git_import.py [hunk 5]
```diff
@@ -50,9 +50,9 @@ class GitImportErrorNoDir(GitImportError):
     def __init__(self, repo_dir):
         super().__init__(
             _(
-                u"Path {0} doesn't exist, please create it, "
-                u"or configure a different path with "
-                u"GIT_REPO_DIR"
+                "Path {0} doesn't exist, please create it, "
```
Comment (HamzaIbnFarooq) [inline edx_sysadmin/git_import.py:53]: The `u"..."` prefixes were removed by black formatter.

**Review by arslanashraf7 — APPROVED** [review]
Works well 👍 .

@HamzaIbnFarooq there is just one thing I would like to ask, Are we up to importing existing mongo logs into MySQL besides this transition? 

cc: @pdpinch

**Comment by blarghmatey** [comment]
I don't see any need to migrate existing logs. They are only useful in the context of an active course and this will be deployed to production after all of the current course runs are complete.

**Pull request merged**